### ISSUE TITLE
Default defined for schema entry overrides the empty set to False

### ISSUE THREAD

**Issue by jjduru** [pyeve/cerberus#462]
Used Cerberus version / latest commit: …

- [ ] I have the capacity to improve the docs when my problem is solved.
- [ ] I have the capacity to submit a patch when a bug is identified.

- [ ] My question does not concern a practical use-case that I can't figure out
      to solve.

----

#### Use-case abstract

A missing option from the document will be caught by the 'empty' set to False and v.error will promptly present a 'required field' error.
However, this behaviour is overridden if the schema also defines a 'default' value, v.errors will show up empty and the validator will return True.

----

#### Bug report / Feature request

Define a validation schema and a document, as dictionaries below. A validation test will return an error, which is the expected behaviour given the 'empty' being set to False:

`# Example #1:
import cerberus
validation_schema = {
    'backup_mount': {
        'required': True, 'empty': False, 'type': 'string'}
    , 'backup_home_sid_spfile': {
        'required': True, 'empty': False, 'type': 'string'}
}
validation_document = {
    'backup_mount': '/backupmount'
}

v = cerberus.validator.Validator({})
v.validate(validation_document, validation_schema)
print(f"Errors: {v.errors}")
`

` ./cerberus.tester.py
{'backup_home_sid_spfile': ['required field']}`

The behaviour of 'empty': False is completely disregarded if we also set a default value in the '
validation_schema', per the example below

`# Example #2:
import cerberus
validation_schema = {
    'backup_mount': {
        'required': True, 'empty': False, 'type': 'string', 'default': 'change_on_install'}
    , 'backup_home_sid_spfile': {
        'required': True, 'empty': False, 'type': 'string',  'default': '%(backup_home_sid)s/spfile'}
}

validation_document = {
    'backup_mount': '/backupmount'
}

v = cerberus.validator.Validator({})
v.validate(validation_document, validation_schema)
print(f"Errors: {v.errors}")`

In the case of example #2, the v.error dictionary is completely empty which is definitely not the expected behaviour. I consider this to be a bug, unless there's some other setting somewhere else I'm not aware of. Thanks,

**Comment by jjduru** [comment]
I see that in case of example #2, I can access the v.normalized(validation_document) as needed. But I still had the expectation to have the v.errors populated and thus I can use it as a signal that something is wrong with the document so it could be corrected.

**Comment by jjduru** [comment]
I do not appreciate you closing the ticket without hearing more reasons. It may work from your perspective but certainly not from mine.

As I said in my previous comment, even if I can collect the v.normalized document and use it, I still need a signal that something was wrong in the original document. That's why v.errors should've been populated in even when I'm using a default value. That's what "'empty': False", to signal that an option does not exist. If you override its behaviour, I might as well not use "'empty': False" anymore because it becomes completely useless.

**Comment by funkyfuture** [comment]
> That's why v.errors should've been populated, even when I'm using a default value per option. That's what "'required': True" is for, to signal that an option does not exist.

no, it exists in the resulting document as you defined a default.

> I might as well not use "'required': True" anymore because it becomes completely useless.

that is correct.

**Comment by jjduru** [comment]
I think you're missing the point: I know that the missing option exists in the normalized document and that I can use it. 

What I said is that v.errors, still needs to be populated even when using 'default', because it can be used as signal that the original document is not formatted correctly. In this case, I could act on it and try to correct it.
In the current state of things, when using a 'default' value per option, I cannot detect that the original document is incorrect, unless I create yet another schema that does not have defaults in it.

And yes, I do expect that if the 'required' option is populated, that v.errors should signal a missing option regardless of 'default' is there or not. That's just good programming.

**Comment by funkyfuture** [comment]
> I think you're missing the point

please consider that you may be missing some point where your false expectations are evoked.

> unless I create yet another schema that does not have defaults in it.

that is surely feasible.

> That's just good programming.

no, it's design. an evolved one.

**Comment by jjduru** [comment]
Looks like this conversation is one-sided and going nowhere. This part particularly is definitely not good design. Have a good day.

**Comment by andreymal** [comment]
This part particularly is definitely good design :)

I also don't understand this problem. If you want `v.errors`, you can just remove a default; if you also want normalized data, it's available at `v.document` even if `v.errors` is not empty (note that invalid fields are still present in `v.document`, but you can easily filter it)

(Actually I think that `v.normalized()` has a bad design)

**Comment by christabone** [comment]
I fail to see how this is bad design? You're asking the schema to provide a default value. The fact that it's an empty field in your document doesn't matter because you're providing a default value in your schema. Hence, there is no error.

It's like getting annoyed at `defaultdict` for creating a new entry instead of throwing a `KeyError`.

**Comment by andreymal** [comment]
Perhaps something like `v.normalized(..., use_default_on_error=True)` may be applicable here

It looks like a good feature request (but not like a bad design of the current implementation)

**Comment by jjduru** [comment]
The schema is there to serve as a "Source of Truth", not a source of data. Is there to signal that option x is missing from the document, and if its' missing, you could use this default value. But the v.errors should still signal the "missing option" part.

I do not need to normalize data in bulk.
